c++/src/kj/compat/http.c++

```diff
@@ -3545,6 +3545,224 @@ namespace {
 
     return parts;
   }
+
+  inline kj::Maybe<kj::Exception> validateTokenTail(const StringPtr s) {
+    // Assert that the extension/parameter token does not include invalid characters between
+    // the end of the valid token and the start of the separator/EOL.
+    //
+    // Suppose END refers to one past the last character we would want to process, for example:
+    //  - if `s` == "client_max_window_bits = 10", then END would be the index of "=".
+    //  - if `s` == "permessage-deflate1234JUNK_VALUES", then END would be s.size()
+    //
+    // We are matching this regex:
+    // ^((client|server)((_max_window_bits\s*(=.*)?)|(_no_context_takeover\s*))$)|(permessage-deflate\s*$)
```

Comment:
Here's a summary since it feels like a lot changed:
1. Given a collection of `kj::String`s (each element is the result of splitting an offer by `;`), we parse into `key[=value]` pairs (this is the `keyMaybeVal` struct + `toKeysAndVals()` function).
2. Next, try to build an `UnverifiedConfig` struct (basically a structured version of param names). This could fail if the structure of the offer is invalid, ex:
    - Too many parameters (we can have 4 at most).
    - Invalid parameter names (we make sure the `key` part of `key[=value]` is valid here.
    - `x_no_context_takeover` parameters have values (i.e. we got `x_no_context_takeover=blah`), when they shouldn't.
    - See the `populateUnverifiedConfig()` function for details.
3. Finally, we convert from an `UnverifiedConfig` struct to a `CompressionParameters` struct. During this process (`validateCompressionConfig()` function) we check the values (if any are present) for the `x_max_window_bits` parameters.

Had to do this in a few parts to avoid merge conflicts :upside_down_face: 

1. Added the structs and functions mentioned above: https://github.com/capnproto/capnproto/pull/1480/commits/c808f9e04cf590fd6fc37220d143b663917bac30
2. Use new functions when parsing as the server: https://github.com/capnproto/capnproto/pull/1480/commits/92c59bd53040e49b86dfef51a13623131381454d
3. Use new functions when parsing as the client: https://github.com/capnproto/capnproto/pull/1480/commits/565bd44992aa488af22291dd5e220de214923108

I'm going to rebase, then delete the _old_ parsing functions and add that fixup below.